src/native/eventpipe/ep-session.c

```diff
@@ -215,6 +219,47 @@ ep_session_alloc (
 	ep_exit_error_handler ();
 }
 
+void
+ep_session_remove_dangling_session_states (EventPipeSession *session)
+{
+	ep_return_void_if_nok (session != NULL);
+
+	EP_RT_DECLARE_LOCAL_THREAD_ARRAY (threads);
+	ep_rt_thread_array_init (&threads);
+
+	ep_thread_get_threads (&threads);
+
+	ep_rt_thread_array_iterator_t threads_iterator = ep_rt_thread_array_iterator_begin (&threads);
+	while (!ep_rt_thread_array_iterator_end (&threads, &threads_iterator)) {
+		EventPipeThread *thread = ep_rt_thread_array_iterator_value (&threads_iterator);
+		if (thread) {
+			EP_SPIN_LOCK_ENTER (&thread->rt_lock, section1);
+				EventPipeThreadSessionState *session_state = ep_thread_get_session_state(thread, session);
+				if (session_state) {
+					// If a buffer tries to write event(s) but never gets a buffer because the maximum total buffer size
+					// has been exceeded, we can leak the EventPipeThreadSessionState* and crash later trying to access 
+					// the session from the thread session state. Whenever we terminate a session we check to make sure
+					// we haven't leaked any thread session states.
+					ep_thread_delete_session_state(thread, session);
```

Comment:
In case where the thread session state ends up in buffer_manager->thread_session_state_list and later deleted when we flush or delete the buffer manager, I guess the order in ep_session_free will prevent us from having any additional copies of thread session state at that point (ep_session_remove_dangling_session_states called after buffer manager and file has been freed)?

Reply:
That's correct, I made sure to add the cleanup code as the last thing before we free the session object so everything except the leaked session states are cleaned up